File: skypydb-js/src/database/mixins/vector/collections/create_collection.ts
import { InputValidator } from "../../../../security/validation";
import { CollectionAlreadyExistsError } from "../../../../errors";
import { CollectionAuditMixin } from "./collection_audit";

export abstract class CreateCollectionMixin extends CollectionAuditMixin {
  create_collection(name: string, metadata?: Record<string, unknown>): void {
    const validated = InputValidator.validate_table_name(name);
    if (this.collection_exists(validated)) {
      throw new CollectionAlreadyExistsError(
        `Collection '${validated}' already exists`,
      );
    }

    const table_name = `vec_${validated}`;
    this.conn
      .prepare(
        `
        CREATE TABLE [${table_name}] (
          id TEXT PRIMARY KEY,
          document TEXT,
          embedding TEXT NOT NULL,
          metadata TEXT,
          created_at TEXT NOT NULL
        )
        `,
      )
      .run();

    this.conn
      .prepare(
        "INSERT INTO _vector_collections (name, metadata, created_at) VALUES (?, ?, ?)",
      )
      .run(validated, JSON.stringify(metadata ?? {}), new Date().toISOString());
  }
}

File: skypydb/database/operations/vector/collections/create_collection.py
"""Collection creation mixin for vector storage."""

import json
from datetime import datetime
from typing import Any, Dict, Optional

from skypydb.security.validation import InputValidator


class CreateCollectionMixin:
    """Provide collection table creation behavior."""

    def create_collection(
        self,
        name: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Create a new collection table and register it in metadata."""

        name = InputValidator.validate_table_name(name)
        if self.collection_exists(name):
            raise ValueError(f"Collection '{name}' already exists")

        table_name = f"vec_{name}"
        cursor = self.conn.cursor()

        cursor.execute(
            f"""
            CREATE TABLE [{table_name}] (
                id TEXT PRIMARY KEY,
                document TEXT,
                embedding TEXT NOT NULL,
                metadata TEXT,
                created_at TEXT NOT NULL
            )
            """
        )
        cursor.execute(
            "INSERT INTO _vector_collections (name, metadata, created_at) VALUES (?, ?, ?)",
            (name, json.dumps(metadata or {}), datetime.now().isoformat()),
        )
        self.conn.commit()
